src/wasm/wasm-type.cpp

```diff
@@ -98,8 +98,8 @@ struct TypeInfo {
       case TupleKind:
         return tuple == other.tuple;
       case RefKind:
-        return ref.heapType == other.ref.heapType &&
-               ref.nullable == other.ref.nullable;
+        // TODO: compare nullability as well, once we properly support that
+        return ref.heapType == other.ref.heapType;
```

Comment:
What made this change necessary? It seems dangerous to me. For one, type equality no longer implies hash equality, and I wouldn't be surprised if that subtly breaks something.

Reply:
without this, the fuzzer runs into issues, because we emit nullable types in places that are not technically valid. but we do still emit a non-nullable type for a `ref.func` for example. and we can end up comparing those.

for now, I think we should just ignore the issue of nullability, which is what this moves us towards.

Reply:
I'm still worried that ignoring issues of nullability by hacking around it whenever it comes up is going to leave us in a real mess where we are totally inconsistent in how we treat nullability. It would be better to ignore non-nullability at its source by either having the parsers translate all non-nullable types to nullable types or making the validator reject all non-nullable types. If we ensure no non-nullable types get into our system, we are free to write code that simply asserts that they are not present. Then, later when we want to support them, we only have to make changes to a few key places and all the asserts will tell us exactly what needs fixing rather than having silently broken code (albeit marked with TODOs) all over the place.

Reply:
That's true, but the total work of now + later would still be the same, I think.

But it sounds like you feel strongly about this? If so I can do the work now.

Note though that the work would not just be changing the parsers, but all other places that generate a type, like a `ref.func` execution in the interpreter, etc.

Reply:
I would appreciate that, thank you! I think it will be much easier to reason about that code with that change.

Reply:
Can we pretend `ref.func` generates a nullable type, and assume/assert that all types are nullable elsewhere? That way, we need to only fix `ref.func` later at the source. By the way, I don't think the interpreter needs to change? (I commented there too) The spec's `ref.func` takes a nullable reference and traps on null anyway.

Reply:
> The spec's ref.func takes a nullable reference and traps on null anyway

`ref.func` => `call_ref`, I think.